File: frontend/src/evaluator.ts
/**
 * Pure rule-evaluation functions: no DOM, no I/O, easy to unit-test.
 *
 * Mirror of `custom_components/smart_icons/evaluator.py`. A divergence
 * between them would mean what the (future) preview UI shows and what
 * the backend injector actually writes disagree.
 *
 * The semantics match DESIGN.md § 4.2:
 *   - thresholds: first matching entry wins; trailing entry with no
 *     comparator is the "else" branch; non-numeric state on a numeric
 *     comparator yields no match for that entry.
 *   - mapping: exact string match on `source.state`; `_else` is the
 *     fallback bucket; missing key + no `_else` → null.
 *
 * (Template mode existed in v0.2 and the v0.3 alpha line but was
 * always inert. Removed entirely in v0.3.0a3.)
 *
 * Per-field merge semantics (v0.3.0a3): `evaluate*` return SPARSE
 * position objects — only fields the matching entry positively
 * addressed appear. `mergeDecorations` layers positions by priority
 * and inflates to the dense {color, icon, background_color} shape.
 *
 * Two kinds of "position":
 *   - positive: a non-empty string value the painter should apply
 *   - release: explicit null / "" / "inherit" / "unset" sentinel —
 *     the rule says "this field should be cleared regardless of what
 *     lower-priority rules contribute."
 *
 * Absence of a field on a rule means "no opinion" and lets lower-
 * priority rules contribute that field.
 */

import type { Decoration, Rule, ThresholdEntry } from './types';

const RELEASE_SENTINELS = new Set(['', 'inherit', 'unset']);
const COMPARATORS = ['lt', 'lte', 'gt', 'gte', 'eq'] as const;
const DECORATION_FIELDS = ['color', 'icon', 'background_color'] as const;

type DecorationField = (typeof DECORATION_FIELDS)[number];

/** Sparse per-field positions a rule takes on a decoration. Only fields
 *  the rule explicitly addresses are present. A value of `null` is an
 *  explicit release (sentinel), distinguished from "field absent."
 *
 *  The merger inflates this to the dense `Decoration` shape the
 *  injector / painter expect; intermediate evaluator returns stay
 *  sparse so the merger can tell "no position" from "release." */
export type DecorationPositions = Partial<Record<DecorationField, string | null>>;

/** Dense decoration emitted by `mergeDecorations`. Same shape as the
 *  v0.2 `Decoration` interface for backward compatibility — every
 *  field is present (null when unaddressed or explicitly released). */
export interface MergedDecoration {
  color: string | null;
  icon: string | null;
  background_color: string | null;
}

function normalizeFieldValue(v: string | null | undefined): string | null {
  if (v == null) return null;
  if (RELEASE_SENTINELS.has(v)) return null;
  return v;
}

/** Build the sparse positions object from a raw decoration dict. Only
 *  keys present in the input become positions; sentinels become null
 *  positions; everything else passes through. */
function normalizeDecoration(
  d: Record<string, unknown> | null | undefined
): DecorationPositions | null {
  if (!d) return null;
  const positions: DecorationPositions = {};
  for (const key of DECORATION_FIELDS) {
    if (!(key in d)) continue;
    positions[key] = normalizeFieldValue(d[key] as string | null | undefined);
  }
  return Object.keys(positions).length > 0 ? positions : null;
}

function entryComparator(t: ThresholdEntry): typeof COMPARATORS[number] | null {
  for (const c of COMPARATORS) {
    if (t[c] !== undefined) return c;
  }
  return null;
}

export function evaluateThresholds(
  thresholds: ThresholdEntry[],
  sourceState: string
): DecorationPositions | null {
  const numericState = Number(sourceState);
  const numericOK = !Number.isNaN(numericState) && sourceState.trim() !== '';

  for (const t of thresholds) {
    const cmp = entryComparator(t);
    let matched = false;

    if (cmp === null) {
      matched = true;
    } else if (cmp === 'eq') {
      const eq = t.eq;
      if (typeof eq === 'number') {
        matched = numericOK && numericState === eq;
      } else {
        matched = sourceState === String(eq);
      }
    } else {
      if (!numericOK) continue;
      switch (cmp) {
        case 'lt':
          matched = numericState < (t.lt as number);
          break;
        case 'lte':
          matched = numericState <= (t.lte as number);
          break;
        case 'gt':
          matched = numericState > (t.gt as number);
          break;
        case 'gte':
          matched = numericState >= (t.gte as number);
          break;
      }
    }

    if (matched) {
      // Pass the entry directly — normalizeDecoration only considers
      // keys that exist on it, so absent decoration fields stay absent
      // (no position) while explicit sentinels become explicit releases.
      return normalizeDecoration(t as Record<string, unknown>);
    }
  }
  return null;
}

export function evaluateMapping(
  mapping: Record<string, Decoration>,
  sourceState: string
): DecorationPositions | null {
  if (Object.prototype.hasOwnProperty.call(mapping, sourceState)) {
    return normalizeDecoration(
      mapping[sourceState] as unknown as Record<string, unknown>
    );
  }
  if (Object.prototype.hasOwnProperty.call(mapping, '_else')) {
    return normalizeDecoration(
      mapping._else as unknown as Record<string, unknown>
    );
  }
  return null;
}

export function evaluateRule(
  rule: Rule,
  sourceState: string | undefined
): DecorationPositions | null {
  if (!rule.enabled) return null;
  if (sourceState == null) return null;

  if (rule.mode === 'thresholds' && rule.thresholds) {
    return evaluateThresholds(rule.thresholds, sourceState);
  }
  if (rule.mode === 'mapping' && rule.mapping) {
    return evaluateMapping(rule.mapping, sourceState);
  }
  return null;
}

/** Layer per-rule positions into a final decoration. Highest-priority
 *  rule that takes a position on a field wins that field; lower-
 *  priority rules fill in fields no higher-priority rule addressed.
 *  Equal priorities: declaration order (first in the input list) wins.
 *
 *  Returns null when no rule positions on anything; otherwise returns
 *  the dense {color, icon, background_color} shape with null in any
 *  unaddressed-or-explicitly-released slot — the painter/injector
 *  treats both equivalently (clear any attribute we'd previously set). */
export function mergeDecorations(
  rules: Rule[],
  positions: (DecorationPositions | null)[]
): MergedDecoration | null {
  // Pair each rule with its index so we can use index as the stable
  // tiebreaker for equal priorities.
  const indexed = rules
    .map((rule, idx) => ({ rule, idx, dec: positions[idx] }))
    .filter((row): row is { rule: Rule; idx: number; dec: DecorationPositions } => row.dec !== null);
  if (indexed.length === 0) return null;
  indexed.sort((a, b) => {
    if (a.rule.priority !== b.rule.priority) return b.rule.priority - a.rule.priority;
    return a.idx - b.idx;
  });

  const merged: Partial<Record<DecorationField, string | null>> = {};
  for (const { dec } of indexed) {
    for (const field of DECORATION_FIELDS) {
      if (field in merged) continue;
      if (field in dec) merged[field] = dec[field] ?? null;
    }
    if (Object.keys(merged).length === DECORATION_FIELDS.length) break;
  }
  if (Object.keys(merged).length === 0) return null;
  return {
    color: merged.color ?? null,
    icon: merged.icon ?? null,
    background_color: merged.background_color ?? null,
  };
}

/** Back-compat alias for v0.2 callers. Same behavior as
 *  mergeDecorations. */
export const pickWinner = mergeDecorations;

File: custom_components/smart_icons/evaluator.py
"""Pure rule-evaluation functions, mirror of frontend/src/evaluator.ts.

Both implementations must stay in sync — backend evaluation runs in the
injector to produce icon + color + background attributes that flow to
HA's frontend naturally, while the TS evaluator is kept for future
panel/preview UI. A divergence between them would mean what the painter
previews and what the integration actually applies disagree.

Semantics follow DESIGN.md § 4.2:
- thresholds: first matching entry wins; an entry with no comparator is
  the "else" branch; non-numeric source state on a numeric comparator
  skips that entry.
- mapping: exact string match; `_else` is the fallback; missing key with
  no `_else` → None.

(Template mode existed in v0.2 and the v0.3 alpha line but was always
inert — runtime evaluation was deferred to "demand-driven." It is
removed entirely in v0.3.0a3 since no real user demand surfaced.)

Per-field merging (v0.3.0a3): when multiple rules target the same
entity, fields are layered by priority. The highest-priority rule that
takes a *position* on a field (color, icon, or background_color) wins
that field. Lower-priority rules can fill in fields the winner did not
address — so a high-priority chip-only rule can coexist with a
low-priority color rule without erasing it. This replaces the v0.2
"winner takes all" rule, which surprised users once `background_color`
joined the decoration set.

Two kinds of "position":
- positive: a non-empty string value the painter should apply
- release: explicit null / "" / "inherit" / "unset" sentinel — the
  rule is saying "this field should be cleared, regardless of what
  lower-priority rules contribute"
Absence of a field on a rule means "I have no opinion on this field"
and lets lower-priority rules contribute.
"""

from __future__ import annotations

from typing import Any

from .const import (
    MAPPING_ELSE_KEY,
    MODE_MAPPING,
    MODE_THRESHOLDS,
)
from .rule import Rule

_RELEASE_SENTINELS = frozenset({"", "inherit", "unset"})
_COMPARATORS = ("lt", "lte", "gt", "gte", "eq")
_DECORATION_FIELDS = ("color", "icon", "background_color")


def _normalize_field(value: Any) -> str | None:
    if value is None:
        return None
    if isinstance(value, str) and value in _RELEASE_SENTINELS:
        return None
    return value


def _normalize_decoration(d: dict[str, Any] | None) -> dict[str, str | None] | None:
    """Return the per-field positions this decoration takes.

    Sparse dict: only fields the input dict mentions appear in the
    output. A field present-but-sentinel maps to None (explicit
    release); a field present-with-value maps to the value; a field
    absent from the input is absent from the output (no position).

    Returns None when the input is empty / takes no positions.
    """
    if not d:
        return None
    positions: dict[str, str | None] = {}
    for key in _DECORATION_FIELDS:
        if key not in d:
            continue
        positions[key] = _normalize_field(d[key])
    return positions or None


def _entry_comparator(entry: dict[str, Any]) -> str | None:
    for cmp in _COMPARATORS:
        if cmp in entry:
            return cmp
    return None


def _coerce_numeric(state: str | None) -> float | None:
    if state is None or state == "":
        return None
    try:
        return float(state)
    except (TypeError, ValueError):
        return None


def evaluate_thresholds(
    thresholds: list[dict[str, Any]], source_state: str
) -> dict[str, str | None] | None:
    numeric = _coerce_numeric(source_state)
    for entry in thresholds:
        cmp = _entry_comparator(entry)
        matched = False
        if cmp is None:
            # Final "else" branch always matches.
            matched = True
        elif cmp == "eq":
            eq = entry["eq"]
            if isinstance(eq, (int, float)):
                matched = numeric is not None and numeric == eq
            else:
                matched = source_state == str(eq)
        else:
            # Numeric comparators require numeric state.
            if numeric is None:
                continue
            threshold = entry[cmp]
            if cmp == "lt":
                matched = numeric < threshold
            elif cmp == "lte":
                matched = numeric <= threshold
            elif cmp == "gt":
                matched = numeric > threshold
            elif cmp == "gte":
                matched = numeric >= threshold
        if matched:
            # Pass the entry directly — _normalize_decoration only
            # considers keys present in the dict, so absent decoration
            # fields stay absent (no position) and explicit sentinels
            # become explicit releases.
            return _normalize_decoration(entry)
    return None


def evaluate_mapping(
    mapping: dict[str, dict[str, Any]], source_state: str
) -> dict[str, str | None] | None:
    if source_state in mapping:
        return _normalize_decoration(mapping[source_state])
    if MAPPING_ELSE_KEY in mapping:
        return _normalize_decoration(mapping[MAPPING_ELSE_KEY])
    return None


def evaluate_rule(
    rule: Rule, source_state: str | None
) -> dict[str, str | None] | None:
    """Return the decoration this rule produces, or None if no match."""
    if not rule.enabled:
        return None
    if source_state is None:
        return None
    if rule.mode == MODE_THRESHOLDS and rule.thresholds:
        return evaluate_thresholds(rule.thresholds, source_state)
    if rule.mode == MODE_MAPPING and rule.mapping:
        return evaluate_mapping(rule.mapping, source_state)
    return None


def merge_decorations(
    rule_decorations: list[tuple[Rule, dict[str, str | None] | None]],
) -> dict[str, str | None] | None:
    """Layer per-rule decoration positions into a single final decoration.

    Highest-priority rule that takes a position on a field wins that
    field. Lower-priority rules fill in fields no higher-priority rule
    addressed. Equal priorities: declaration order (first occurrence)
    wins, matching the TS evaluator and the v0.2 tiebreaker.

    Returns a dict with all three decoration fields (None for fields no
    rule positioned on, OR fields where the winning position was a
    sentinel release). The injector treats both the same — pop the
    attribute if it's no longer set. Returns None only when no rule
    positioned on anything, so callers can distinguish "no decoration"
    from "decoration explicitly clears the host."
    """
    indexed = [
        (rule.priority, idx, dec)
        for idx, (rule, dec) in enumerate(rule_decorations)
        if dec is not None
    ]
    if not indexed:
        return None
    # Sort: higher priority first; for ties, earlier declaration index
    # first (negate priority so the natural ascending sort gives us
    # high-priority-first).
    indexed.sort(key=lambda t: (-t[0], t[1]))

    merged: dict[str, str | None] = {}
    for _prio, _idx, dec in indexed:
        for field in _DECORATION_FIELDS:
            if field in merged:
                continue  # already taken by a higher-priority rule
            if field in dec:
                merged[field] = dec[field]
        if len(merged) == len(_DECORATION_FIELDS):
            break  # every field claimed; no need to keep walking

    if not merged:
        # No rule took any field-level position (shouldn't happen
        # because _normalize_decoration returns None when sparse, but
        # defend against future shape changes).
        return None
    # Inflate to the dense shape the injector expects. Fields not
    # claimed by any rule remain None — same as "field released."
    return {field: merged.get(field) for field in _DECORATION_FIELDS}


# Back-compat alias. Old callers can switch incrementally.
pick_winner = merge_decorations
